File: nn_classifier.m
classdef nn_classifier

    properties
        examples = {};
        mean_ = [];
        cov_ = [];
    end

    methods
        function self = train(self, sample_labels, samples)
            self.examples = {};
            self.mean_ = mean(samples);
            self.cov_ = cov(samples);
            for i = 1:size(samples, 1)
                self.examples{end+1} = {sample_labels{i}, (samples(i, :) - self.mean_)./sqrt(diag(self.cov_)')};
            end
        end

        function prediction = predict(self, x)
            best = {inf, ''};
            x = (x - self.mean_)./sqrt(diag(self.cov_)');
            for i = 1:size(self.examples, 2)
                d = norm(x - self.examples{i}{2});
                if d < best{1}
                    best = {d, self.examples{i}{1}};
                end
            end
            prediction = best{2};
        end
    end
end

File: nn_classifier.py
#!/usr/bin/python

from __future__ import print_function, division
from numpy.linalg import norm
from numpy import *

class nn_classifier(object):

    def __init__(self, dim, k=1):
        self._examples = None;
        self._mean = None;
        self._cov= None;
        self._dim = dim;
        self._k = k;

    @property
    def dim(self):
        return self._dim;

    @property
    def examples(self):
        return self._examples;

    def train(self, classes):
        samples = [];
        for k in classes: samples += classes[k];
        self._mean = mean(samples, axis=0);
        self._cov = cov(transpose(samples));
        if not self._cov.shape:
            self._cov.shape = (1, 1);
        self._examples = ();
        for label in classes:
            for example in classes[label]:
                self._examples = self._examples + (((example - self._mean)/sqrt(diag(self._cov)), label),);

    def predict(self, x):
        x = (x - self._mean)/sqrt(diag(self._cov));
        return min(self.examples, key = lambda e: norm(e[0] - x))[1];
